database/src/ops/tx.rs

```diff
@@ -1,64 +1,296 @@
 //! Transactions.
 
 //---------------------------------------------------------------------------------------------------- Import
+use bytemuck::TransparentWrapper;
 
-//---------------------------------------------------------------------------------------------------- Free Functions
-/// TODO
-pub fn add_transaction() {
-    todo!()
-}
+use monero_pruning::PruningSeed;
+use monero_serai::transaction::{Timelock, Transaction};
 
-/// TODO
-pub fn add_transaction_data() {
-    todo!()
-}
+use cuprate_types::{OutputOnChain, TransactionVerificationData, VerifiedBlockInformation};
 
-/// TODO
-pub fn remove_transaction() {
-    todo!()
-}
+use crate::{
+    database::{DatabaseIter, DatabaseRo, DatabaseRw},
+    env::EnvInner,
+    error::RuntimeError,
+    ops::{
+        blockchain::chain_height,
+        macros::{doc_add_block_inner_invariant, doc_error},
+        property::get_blockchain_pruning_seed,
+    },
+    tables::{
+        BlockBlobs, BlockHeights, BlockInfos, KeyImages, NumOutputs, Outputs, PrunableHashes,
+        PrunableTxBlobs, PrunedTxBlobs, RctOutputs, Tables, TablesMut, TxBlobs, TxHeights, TxIds,
+        TxUnlockTime,
+    },
+    transaction::{TxRo, TxRw},
+    types::{
+        AmountIndices, BlockHash, BlockHeight, BlockInfo, KeyImage, Output, PreRctOutputId,
+        RctOutput, TxBlob, TxHash, TxId,
+    },
+    StorableVec,
+};
 
-/// TODO
-pub fn remove_transaction_data() {
-    todo!()
-}
+//---------------------------------------------------------------------------------------------------- Private
+/// Add a [`TransactionVerificationData`] to the database.
+///
+/// The `block_height` is the block that this `tx` belongs to.
+///
+/// Note that the caller's input is trusted implicitly and no checks
+/// are done (in this function) whether the `block_height` is correct or not.
+///
+/// # Notes
+/// This uses the [`chain_height`] when addings to the [`TxHeights`] table.
+///
+#[doc = doc_add_block_inner_invariant!()]
+#[doc = doc_error!()]
+#[inline]
+pub fn add_tx(
```

Comment:
I think I would prefer if output data was added in this function, it feels more at home here than `add_block`

Reply:
Sure, but I'll note that this breaks the logical `data type <-> fn` groupings, i.e:
- `Transaction` <-> `{get,remove}_tx()`
- `Output` <-> `{get,remove}_output()`

As of now each sub-function only mutates its related tables. This makes `add_tx()` have a side effect that it also touches outputs.